src/libcore/num/mod.rs

```diff
@@ -1446,21 +1447,29 @@ doit! { i8 i16 i32 i64 isize u8 u16 u32 u64 usize }
 
 fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32)
                                          -> Result<T, ParseIntError> {
+    const MINUS: u8 = '-' as u8;
     use self::IntErrorKind::*;
     use self::ParseIntError as PIE;
+
     assert!(radix >= 2 && radix <= 36,
            "from_str_radix_int: must lie in the range `[2, 36]` - found {}",
            radix);
 
+    if src.is_empty() {
+        return Err(PIE { kind: Empty });
+    }
+
     let is_signed_ty = T::from_u32(0) > T::min_value();
 
-    match src.slice_shift_char() {
-        Some(('-', "")) => Err(PIE { kind: Empty }),
-        Some(('-', src)) if is_signed_ty => {
+    // all valid digits are ascii
+    let src = src.as_bytes();
+    match (src[0], &src[1..])  {
+        (MINUS, src) if src.is_empty() => Err(PIE { kind: Empty }),
+        (MINUS, src) if is_signed_ty => {
             // The number is negative
             let mut result = T::from_u32(0);
-            for c in src.chars() {
-                let x = match c.to_digit(radix) {
+            for &c in src {
+                let x = match (c as char).to_digit(radix) {
```

Comment:
Please write a wrapper function with a comment to perform the operation `(c as char).to_digit(radix)`.  The fact that `char::to_digit` returns None for Unicode scalar values between 128 and 255 isn't immediately obvious.